src/libtiled/mapwriter.cpp

```diff
@@ -426,11 +428,34 @@ void MapWriterPrivate::writeTileLayer(QXmlStreamWriter &w,
     if (!compression.isEmpty())
         w.writeAttribute(QLatin1String("compression"), compression);
 
+    if (mMap->singleSheet())
+    {
+        bool tSetFound = false;
+        QString tname;
+        for (int y = 0; y < tileLayer->height(); ++y) {
+            for (int x = 0; x < tileLayer->width(); ++x) {
+                const Cell c = tileLayer->cellAt(x, y);
+                if (!c.isEmpty()) {
+                    tname = c.tile->tileset()->name();
+                    tSetFound = true;
+                }
+            }
+        }
+        if (tSetFound)
+            w.writeAttribute(QLatin1String("tileset"), QLatin1String(tname.toLatin1()));
+    }
+
     if (mLayerDataFormat == Map::XML) {
         for (int y = 0; y < tileLayer->height(); ++y) {
             for (int x = 0; x < tileLayer->width(); ++x) {
-                const unsigned gid = mGidMapper.cellToGid(tileLayer->cellAt(x, y));
+                const Cell c = tileLayer->cellAt(x, y);
+                unsigned gid = mGidMapper.cellToGid(c);
                 w.writeStartElement(QLatin1String("tile"));
+                if (mMap->singleSheet())
+                    if (c.isEmpty())
+                        gid = 0;
+                    else
+                        gid = c.tile->id();
```

Comment:
So in this mode, it is in fact not possible for a layer to have "empty" cells, since they are indistinguishable from the first tile in the tileset.